flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisConsumer.java

```diff
@@ -267,38 +293,84 @@ public void close() throws Exception {
 	// ------------------------------------------------------------------------
 
 	@Override
-	public HashMap<KinesisStreamShard, SequenceNumber> snapshotState(long checkpointId, long checkpointTimestamp) throws Exception {
+	public void snapshotState(FunctionSnapshotContext context) throws Exception {
 		if (lastStateSnapshot == null) {
 			LOG.debug("snapshotState() requested on not yet opened source; returning null.");
-			return null;
-		}
-
-		if (fetcher == null) {
+		} else if (fetcher == null) {
 			LOG.debug("snapshotState() requested on not yet running source; returning null.");
-			return null;
-		}
-
-		if (!running) {
+		} else if (!running) {
 			LOG.debug("snapshotState() called on closed source; returning null.");
-			return null;
+		} else {
+			if (LOG.isDebugEnabled()) {
+				LOG.debug("Snapshotting state ...");
+			}
+
+			offsetsStateForCheckpoint.clear();
+			lastStateSnapshot = fetcher.snapshotState();
+
+			if (LOG.isDebugEnabled()) {
+				LOG.debug("Snapshotted state, last processed sequence numbers: {}, checkpoint id: {}, timestamp: {}",
+					lastStateSnapshot.toString(), context.getCheckpointId(), context.getCheckpointTimestamp());
+			}
+
+			for (Map.Entry<KinesisStreamShard, SequenceNumber> entry : lastStateSnapshot.entrySet()) {
+				offsetsStateForCheckpoint.add(Tuple2.of(entry.getKey(), entry.getValue()));
+			}
 		}
+	}
 
-		if (LOG.isDebugEnabled()) {
-			LOG.debug("Snapshotting state ...");
+	@Override
+	public void initializeState(FunctionInitializationContext context) throws Exception {
+		TypeInformation<Tuple2<KinesisStreamShard, SequenceNumber>> tuple = new TupleTypeInfo<>(
+			TypeInformation.of(KinesisStreamShard.class),
+			TypeInformation.of(SequenceNumber.class)
+		);
+
+		offsetsStateForCheckpoint = context.getOperatorStateStore().getUnionListState(
+			new ListStateDescriptor<>(DefaultOperatorStateBackend.DEFAULT_OPERATOR_STATE_NAME, tuple));
+
+		if (context.isRestored()) {
+			if (sequenceNumsToRestore == null) {
+				sequenceNumsToRestore = new HashMap<>();
+				for (Tuple2<KinesisStreamShard, SequenceNumber> kinesisOffset : offsetsStateForCheckpoint.get()) {
+					sequenceNumsToRestore.put(kinesisOffset.f0, kinesisOffset.f1);
+				}
+
+				LOG.info("Setting restore state in the FlinkKinesisConsumer.");
+				if (LOG.isDebugEnabled()) {
+					LOG.debug("Using the following offsets: {}", sequenceNumsToRestore);
+				}
+			} else if (sequenceNumsToRestore.isEmpty()) {
+				sequenceNumsToRestore = null;
+			}
+		} else {
+			LOG.info("No restore state for FlinkKinesisConsumer.");
 		}
+	}
 
-		lastStateSnapshot = fetcher.snapshotState();
+	@Override
+	public void restoreState(HashMap<KinesisStreamShard, SequenceNumber> restoredState) throws Exception {
+		LOG.info("{} (taskIdx={}) restoring offsets from an older version.",
```

Comment:
I think the other log messages follow a different format then this.
The others do something like `Subtask {} is restoring offset from an older version`?